src/Stratis.Bitcoin/Base/ConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,434 @@
+
+        /// <summary>List of block headers that were not yet consumed by <see cref="ConsensusManager"/>.</summary>
+        /// <remarks>Should be protected by <see cref="asyncLock"/>.</remarks>
+        private readonly List<BlockHeader> cachedHeaders;
+
+        /// <summary>Protects access to <see cref="cachedHeaders"/>.</summary>
+        private readonly AsyncLock asyncLock;
+
+        public ConsensusManagerBehavior(ConcurrentChain chain, IInitialBlockDownloadState initialBlockDownloadState, ConsensusManager consensusManager, IPeerBanning peerBanning, IConnectionManager connectionManager, ILoggerFactory loggerFactory)
+        {
+            this.loggerFactory = loggerFactory;
+            this.initialBlockDownloadState = initialBlockDownloadState;
+            this.consensusManager = consensusManager;
+            this.chain = chain;
+            this.connectionManager = connectionManager;
+            this.peerBanning = peerBanning;
+
+            this.cachedHeaders = new List<BlockHeader>();
+            this.asyncLock = new AsyncLock();
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        ///  <inheritdoc />
+        protected override void AttachCore()
+        {
+            this.logger.LogTrace("()");
+
+            // Initialize auto sync timer.
+            this.autosyncTimer = new Timer(async (o) =>
+            {
+                this.logger.LogTrace("()");
+
+                await this.TrySyncAsync(true).ConfigureAwait(false);
+
+                this.logger.LogTrace("(-)");
+            }, null, 0, (int)TimeSpan.FromMinutes(AutosyncIntervalMinutes).TotalMilliseconds);
+
+            if (this.AttachedPeer.State == NetworkPeerState.Connected)
+                this.AttachedPeer.MyVersion.StartHeight = this.consensusManager.Tip.Height;
+
+            this.AttachedPeer.StateChanged.Register(this.OnStateChangedAsync);
+            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceivedAsync, true);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Presents cached headers to <see cref="ConsensusManager"/> from the cache if any and removes consumed from the cache.</summary>
+        /// <param name="newTip">New consensus tip.</param>
+        public async Task<ConnectNewHeadersResult> ConsensusTipChangedAsync(ChainedHeader newTip)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(newTip), newTip);
+
+            ConnectNewHeadersResult result = null;
+            bool syncRequired = false;
+
+            using (await this.asyncLock.LockAsync().ConfigureAwait(false))
+            {
+                if (this.cachedHeaders.Count != 0)
+                {
+                    try
+                    {
+                        result = await this.PresentHeadersLockedAsync(this.cachedHeaders, false).ConfigureAwait(false);
+
+                        if (result != null)
```

Comment:
The same here:

```
 if (result != null)
    return;
```

Reply:
Used to be like that, @Aprogiena insisted on doing it the way it is right now. I don't mind.

Reply:
Probably he meant `result == null`?

Reply:
I can change it in order to return if the result is null. @Aprogiena are you ok with this?

I remember that's the place where I used to have it but then we changed it earlier today.

Reply:
The original you had wasn't `== null`, I would be okay with `== null`.

Reply:
Fixed